### ISSUE TITLE
Customize automatic re-run logic

### ISSUE THREAD

**Issue by fkorotkov** [cirruslabs/cirrus-ci-docs#238]
# Description

It will be nice to have a customizable re-run policy supporting fancy expressions like following:

```yaml
task:
  rerun: "status('build') == 'failed'"
```
or even:

```yaml
task:
  rerun: "'TimeoutException' in logs('test')"
```

**Comment by saberduck** [comment]
hello @fkorotkov , 

any plan to make this available? we have some jobs intermittently timeouting, this would be very useful.

**Comment by fkorotkov** [comment]
@saberduck do you build Cirrus' timeouts (the default 1 hour one, for example)? Or your build system timeouts to do something?

I'm not 100% sure about the final design here though. I think it seems that the most consis solution will be something like:

```
rerun_when_log_contains: 
  - Timeout
  - timeout
```

My initial thoughts about expressions seems unnecessary here and probably can be a part of #496

**Comment by pavel-mikula-sonarsource** [comment]
Hi @fkorotkov,

We have random timeouts on some intermediate steps like cloning repo, downloading nuget packages, downloading code analyzers, uploading code analysis results.

Some of those timeouts doesn't contain word `timeout` but stuff like `bad gateway`, `internal server error` (actually meaning a timeout), `IllegalStateException: Fail to request https://...` so it would be great to be able to configure it somehow.

**Comment by fkorotkov** [comment]
Fixed via #858 

See documentation on the website https://cirrus-ci.org/guide/programming-tasks/#hooks
